tools/make_patch.c

```diff
@@ -0,0 +1,437 @@
+	return address < 0x8000 ? 0x4000 // ROM
+	     : address < 0xa000 ? 0xa000 // VRAM
+	     : address < 0xc000 ? 0xc000 // SRAM
+	     : address < 0xd000 ? 0xd000 // WRAM
+	     : 0xffff; // HRAM
+}
+
+void symbol_append(struct Symbol **symbols, const char *name, int bank, int address) {
+	size_t name_len = strlen(name) + 1;
+	struct Symbol *symbol = xmalloc(sizeof(*symbol) + name_len);
+	symbol->address = address;
+	symbol->offset = bank > 0 ? address + (bank - 1) * get_address_type_limit(address) : address;
+	memcpy(symbol->name, name, name_len);
+	symbol->next = *symbols;
+	*symbols = symbol;
+}
+
+void symbol_free(struct Symbol *symbols) {
+	for (struct Symbol *next; symbols; symbols = next) {
+		next = symbols->next;
+		free(symbols);
+	}
+}
+
+const struct Symbol *symbol_find(const struct Symbol *symbols, const char *name) {
+	size_t name_len = strlen(name);
+	for (const struct Symbol *symbol = symbols; symbol; symbol = symbol->next) {
+		size_t sym_name_len = strlen(symbol->name);
+		if (name_len > sym_name_len) {
+			continue;
+		}
+		const char *sym_name = symbol->name;
+		if (name[0] == '.') {
+			// If `name` is a local label, compare it to the local part of `symbol->name`
+			sym_name += sym_name_len - name_len;
+		}
+		if (!strcmp(sym_name, name)) {
+			return symbol;
+		}
+	}
+	error_exit("Error: Unknown symbol: \"%s\"\n", name);
+}
+
+const struct Symbol *symbol_find_cat(const struct Symbol *symbols, const char *prefix, const char *suffix) {
+	char *sym_name = xmalloc(strlen(prefix) + strlen(suffix) + 1);
+	sprintf(sym_name, "%s%s", prefix, suffix);
+	const struct Symbol *symbol = symbol_find(symbols, sym_name);
+	free(sym_name);
+	return symbol;
+}
+
+int parse_number(const char *input, int base) {
+	char *endptr;
+	int n = strtol(input, &endptr, base);
+	if (endptr == input || *endptr || n < 0) {
+		error_exit("Error: Cannot parse number: \"%s\"", input);
+	}
+	return n;
+}
+
+void parse_symbol_value(char *input, int *restrict bank, int *restrict address) {
+	char *colon = strchr(input, ':');
+	if (!colon) {
+		error_exit("Error: Cannot parse bank+address: \"%s\"", input);
+	}
+	*colon++ = '\0';
+	*bank = parse_number(input, 16);
+	*address = parse_number(colon, 16);
+}
+
+void parse_symbols(const char *filename, struct Symbol **symbols) {
+	FILE *file = xfopen(filename, 'r');
+	struct Buffer *buffer = buffer_create(1);
+
+	enum { SYM_PRE, SYM_VALUE, SYM_SPACE, SYM_NAME } state = SYM_PRE;
+	int bank = 0;
+	int address = 0;
+
+	for (;;) {
+		int c = getc(file);
+		if (c == EOF || c == '\n' || c == '\r' || c == ';' || (state == SYM_NAME && (c == ' ' || c == '\t'))) {
+			if (state == SYM_NAME) {
+				// The symbol name has ended; append the buffered symbol
+				buffer_append(buffer, &(char []){'\0'});
+				symbol_append(symbols, buffer->data, bank, address);
+			}
+			// Skip to the next line, ignoring anything after the symbol value and name
+			state = SYM_PRE;
+			for (; c != EOF && c != '\n' && c != '\r'; c = getc(file));
```

Comment:
(opinion)
`while (c != EOF && ...) c = getc(file);` (or expand with brackets if you really have to)

Reply:
Aw, fine. :P Same for the other `for (; condition; increment); // no body`.